### PR TITLE
fix: delete dataset columns and metrics on the REST API

### PR DESCRIPTION
**Pull request by dpgaspar**
### SUMMARY
Add the ability to delete columns and metrics using the REST API PUT method for datasets, also add 2 new REST endpoints on the dataset namespace.

`api/v1/dataset/<id>/column/<id>` HTTP DELETE
`api/v1/dataset/<id>/metric/<id>` HTTP DELETE

Add 

The columns and metrics are added, changed or deleted in one go defined on the **PUT payload**:
Example on how to add a column on `energy_usage`:
```
{
  "columns": [
     {
        "column_name": "source",
        "description": null,
        "expression": "",
        "filterable": true,
        "groupby": true,
        "id": 1,
        "is_active": true,
        "is_dttm": false,
        "python_date_format": null,
        "type": "VARCHAR(255)",
        "uuid": "56fcb12f-83bf-4eed-994d-e2dc888922d9",
        "verbose_name": null
      },
      {
        "column_name": "target",
        "description": null,
        "expression": "",
        "filterable": true,
        "groupby": true,
        "id": 2,
        "is_active": true,
        "is_dttm": false,
        "python_date_format": null,
        "type": "VARCHAR(255)",
        "uuid": "8f2385b7-30a2-408e-a792-ea613c3889d9",
        "verbose_name": null
      },
      {
        "column_name": "value",
        "description": null,
        "expression": "",
        "filterable": true,
        "groupby": true,
        "id": 3,
        "is_active": true,
        "is_dttm": false,
        "python_date_format": null,
        "type": "DOUBLE PRECISION",
        "uuid": "fbaf8c58-ae5f-4791-afc3-cc73cd0c26a3",
        "verbose_name": null
      },
      {"column_name": "NEW_COLUMN", "type": "INTEGER"}
]
}
```

Removing the `NEW_COLUMN` column:
```
  "columns": [
     {
        "column_name": "source",
        "description": null,
        "expression": "",
        "filterable": true,
        "groupby": true,
        "id": 1,
        "is_active": true,
        "is_dttm": false,
        "python_date_format": null,
        "type": "VARCHAR(255)",
        "uuid": "56fcb12f-83bf-4eed-994d-e2dc888922d9",
        "verbose_name": null
      },
      {
        "column_name": "target",
        "description": null,
        "expression": "",
        "filterable": true,
        "groupby": true,
        "id": 2,
        "is_active": true,
        "is_dttm": false,
        "python_date_format": null,
        "type": "VARCHAR(255)",
        "uuid": "8f2385b7-30a2-408e-a792-ea613c3889d9",
        "verbose_name": null
      },
      {
        "column_name": "value",
        "description": null,
        "expression": "",
        "filterable": true,
        "groupby": true,
        "id": 3,
        "is_active": true,
        "is_dttm": false,
        "python_date_format": null,
        "type": "DOUBLE PRECISION",
        "uuid": "fbaf8c58-ae5f-4791-afc3-cc73cd0c26a3",
        "verbose_name": null
      }
]
}
```

Updating a column:
```
  "columns": [
     {
        "column_name": "source",
        "description": null,
        "expression": "",
        "filterable": true,
        "groupby": true,
        "id": 1,
        "is_active": true,
        "is_dttm": false,
        "python_date_format": null,
        "type": "VARCHAR(255)",
        "uuid": "56fcb12f-83bf-4eed-994d-e2dc888922d9",
        "verbose_name": null
      },
      {
        "column_name": "target",
        "description": null,
        "expression": "",
        "filterable": true,
        "groupby": true,
        "id": 2,
        "is_active": true,
        "is_dttm": false,
        "python_date_format": null,
        "type": "VARCHAR(255)",
        "uuid": "8f2385b7-30a2-408e-a792-ea613c3889d9",
        "verbose_name": null
      },
      {
        "column_name": "value",
        "description": null,
        "expression": "",
        "filterable": true,
        "groupby": true,
        "id": 3,
        "is_active": true,
        "is_dttm": false,
        "python_date_format": null,
        "type": "DOUBLE PRECISION",
        "uuid": "fbaf8c58-ae5f-4791-afc3-cc73cd0c26a3",
        "verbose_name": null
      },
      {id: "911", "column_name": "NEW_COLUMN_CHANGED", "type": "INTEGER"}
]
}
```

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/datasets/dao.py [hunk 1]
```diff
@@ -186,9 +169,77 @@ def update(  # pylint: disable=W:279
...
+            if column_id:
+                column_obj = db.session.query(TableColumn).get(column_id)
+                column_obj = DatasetDAO.update_column(column_obj, column, commit=commit)
+            else:
+                column_obj = DatasetDAO.create_column(column, commit=commit)
+            new_columns.append(column_obj)
+        # Checks if an exiting column is missing from properties and delete it
+        for existing_column in model.columns:
+            if existing_column.column_name not in [
+                column["column_name"] for column in property_columns
```
Comment (willbarrett) [inline superset/datasets/dao.py:205]: Can name be changed? Typically I'd expect this check to happen on the ID level.

File: superset/datasets/dao.py [hunk 2]
```diff
@@ -186,9 +169,77 @@ def update(  # pylint: disable=W:279
...
+                metric_obj = db.session.query(SqlMetric).get(metric_id)
+                metric_obj = DatasetDAO.update_metric(metric_obj, metric, commit=commit)
+            else:
+                metric_obj = DatasetDAO.create_metric(metric, commit=commit)
+            new_metrics.append(metric_obj)
+
+        # Checks if an exiting column is missing from properties and delete it
+        for existing_metric in model.metrics:
+            if existing_metric.metric_name not in [
+                metric["metric_name"] for metric in property_metrics
```
Comment (willbarrett) [inline superset/datasets/dao.py:239]: Same question here - I would expect us to use `id`

File: superset/datasets/dao.py (hunk 1 above)
Comment (dpgaspar) [reply superset/datasets/dao.py]: good catch, changed

**Comment by etr2460** [comment]
Hmmm, does this mean that to only change one column/metric you need to send a payload with all the other columns/metrics and all their attributes along with it? That's not ideal since we have a dataset with over 10k metrics and 5k columns.

Could we instead add the api directly onto the SqlMetric and SqlColumn models? I think it makes sense to create, update, and delete these outside the scope of the Dataset, as long as when creating them you add the dataset ID

**Comment by dpgaspar** [comment]
> Hmmm, does this mean that to only change one column/metric you need to send a payload with all the other columns/metrics and all their attributes along with it? That's not ideal since we have a dataset with over 10k metrics and 5k columns.
> 
> Could we instead add the api directly onto the SqlMetric and SqlColumn models? I think it makes sense to create, update, and delete these outside the scope of the Dataset, as long as when creating them you add the dataset ID

I understand, took the decision because the current process works with a full payload and the frontend modal (although it targets a different endpoint) has the same logic. Do you notice any problems with it?

I'll add 2 additional endpoints to delete:
 `/api/v1/datasets/<id>/metric/<id>` HTTP DELETE
`/api/v1/datasets/<id>/column/<id>` HTTP DELETE

**Comment by etr2460** [comment]
I guess the only concern I have is that although the current process works with a full payload, that doesn't seem very efficient. And this API change would then force you to always send the full payload back everytime. Should we consider decoupling CRUD actions on columns and metrics from the dataset with the new API? The frontend modal could continue to use the same logic on the different endpoint and eventually we could move to a decoupled API that's more efficient

**Comment by dpgaspar** [comment]
@etr2460 just updated this with a couple of separate endpoints

**Comment by etr2460** [comment]
The separate endpoints look great! Not sure if the approach for PUT requests to datasets to require the full dataset (including columns and metrics) is the best option, but since it works for now we can cross that bridge when we get to it

File: superset/datasets/columns/commands/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+class DatasetColumnNotFoundError(CommandException):
+    message = _("Dataset column not found.")
+
+
+class DatasetColumnDeleteFailedError(CommandException):
+    message = _("Dataset column delete failed.")
+
+
+class DatasetColumnForbiddenError(CommandException):
+    message = _("Changing this dataset is forbidden")
```
Comment (etr2460) [inline superset/datasets/columns/commands/exceptions.py:31]: nit: add a period

File: superset/datasets/metrics/commands/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+class DatasetMetricNotFoundError(CommandException):
+    message = _("Dataset metric not found.")
+
+
+class DatasetMetricDeleteFailedError(CommandException):
+    message = _("Dataset metric delete failed.")
+
+
+class DatasetMetricForbiddenError(CommandException):
+    message = _("Changing this dataset is forbidden")
```
Comment (etr2460) [inline superset/datasets/metrics/commands/exceptions.py:31]: same nit here

**Review by etr2460 — APPROVED** [review]

**Pull request merged**